crates/shared/src/sources/balancer_v2/pool_fetching.rs

```diff
@@ -199,74 +194,63 @@ impl BalancerPoolFetching for BalancerPoolFetcher {
         token_pairs: HashSet<TokenPair>,
         at_block: Block,
     ) -> Result<FetchedBalancerPools> {
-        let pool_ids = self
-            .pool_registry
-            .pool_ids_for_token_pairs(&token_pairs)
-            .await;
-        let fetched_stable_pools = self
-            .stable_pool_reserve_cache
-            .fetch(pool_ids.clone(), at_block)
-            .await?;
-        let fetched_weighted_pools = self
-            .weighted_pool_reserve_cache
-            .fetch(pool_ids, at_block)
-            .await?;
-        // Return only those pools which are not paused.
-        Ok(FetchedBalancerPools {
-            stable_pools: filter_paused(fetched_stable_pools),
-            weighted_pools: filter_paused(fetched_weighted_pools),
-        })
-    }
-}
+        let pool_ids = self.fetcher.pool_ids_for_token_pairs(token_pairs).await;
+        let pools = self.fetcher.pools_by_id(pool_ids, at_block).await?;
 
-fn filter_paused<T: BalancerPoolEvaluating>(pools: Vec<T>) -> Vec<T> {
-    pools
-        .into_iter()
-        .filter(|pool| !pool.properties().paused)
-        .collect()
+        // For now, split the `Vec<Pool>` into a `FetchedBalancerPools` to keep
+        // compatibility with the rest of the project. This should eventually
+        // be removed and we should use `balancer_v2::pools::Pool` everywhere
+        // instead.
+        let fetched_pools = pools.into_iter().fold(
+            FetchedBalancerPools::default(),
+            |mut fetched_pools, pool| {
+                match pool.kind {
+                    PoolKind::Weighted(state) => fetched_pools
+                        .weighted_pools
+                        .push(WeightedPool::new_unpaused(pool.id, state)),
+                    PoolKind::Stable(state) => fetched_pools
+                        .stable_pools
+                        .push(StablePool::new_unpaused(pool.id, state)),
+                }
+                fetched_pools
+            },
+        );
+
+        Ok(fetched_pools)
+    }
 }
 
 #[async_trait::async_trait]
 impl Maintaining for BalancerPoolFetcher {
     async fn run_maintenance(&self) -> Result<()> {
-        futures::try_join!(
-            self.pool_registry.run_maintenance(),
-            self.stable_pool_reserve_cache.update_cache(),
-            self.weighted_pool_reserve_cache.update_cache(),
-        )?;
-        Ok(())
+        self.fetcher.run_maintenance().await
     }
 }
 
+/// Extract the pool address from an ID.
+///
+/// This takes advantage that the first 20 bytes of the ID is the address of
+/// the pool. For example the GNO-BAL pool with ID
+/// `0x36128d5436d2d70cab39c9af9cce146c38554ff0000200000000000000000009`:
+/// <https://etherscan.io/address/0x36128D5436d2d70cab39C9AF9CcE146C38554ff0>
+fn pool_address_from_id(pool_id: H256) -> H160 {
+    let mut address = H160::default();
```

Comment:
nit: Maybe a bit more succinct `H160::from_slice(&pool_id[..20])` would do the job?

Reply:
Would this function be useful maybe for other liquidity sources?

Reply:
This is a Balancer specific detail.